workflows/src/main/kotlin/com/r3/corda/lib/ci/SyncKeyMappingFlow.kt

```diff
@@ -0,0 +1,105 @@
+        val requestedIdentities = session.sendAndReceive<List<AbstractParty>>(confidentialIdentities).unwrap { req ->
+            require(req.all { it in confidentialIdentities }) {
+                "${session.counterparty} requested a confidential identity not part of transaction: ${tx.id}"
+            }
+            req
+        }
+
+        val resolvedIds = requestedIdentities.map { serviceHub.identityService.wellKnownPartyFromAnonymous(it) }.filter { it != null }
+        session.send(resolvedIds)
+    }
+
+    private fun extractConfidentialIdentities(): List<AbstractParty> {
+        val inputStates: List<ContractState> = (tx.inputs.toSet()).mapNotNull {
+            try {
+                serviceHub.loadState(it).data
+            } catch (e: TransactionResolutionException) {
+                null
+            }
+        }
+        val states: List<ContractState> = inputStates + tx.outputs.map { it.data }
+        val identities: Set<AbstractParty> = states.flatMap(ContractState::participants).toSet()
+
+        return identities
+                .filter { serviceHub.networkMapCache.getNodesByLegalIdentityKey(it.owningKey).isEmpty() }
+                .toList()
+    }
+}
+
+class SyncKeyMappingFlowHandler(private val otherSession: FlowSession) : FlowLogic<Boolean>() {
+    companion object {
+        object RECEIVING_IDENTITIES : ProgressTracker.Step("Receiving confidential identities.")
+        object RECEIVING_PARTIES : ProgressTracker.Step("Receiving potential party objects for unknown identities.")
+        object NO_PARTIES_RECEIVED : ProgressTracker.Step("None of the requested unknown parties were resolved by the counter party. " +
+                "Terminating the flow early.")
+        object REQUESTING_PROOF_OF_ID : ProgressTracker.Step("Requesting a signed key to party mapping for the received parties to verify" +
+                "the authenticity of the party.")
+        object IDENTITIES_SYNCHRONISED : ProgressTracker.Step("Identities have finished synchronising.")
+    }
+
+    override val progressTracker: ProgressTracker = ProgressTracker(RECEIVING_IDENTITIES, RECEIVING_PARTIES, NO_PARTIES_RECEIVED, REQUESTING_PROOF_OF_ID, IDENTITIES_SYNCHRONISED)
+
+    @Suspendable
+    override fun call(): Boolean {
+        progressTracker.currentStep = RECEIVING_IDENTITIES
+        val allConfidentialIds = otherSession.receive<List<AbstractParty>>().unwrap { it }
+        val unknownIdentities = allConfidentialIds.filter { serviceHub.identityService.wellKnownPartyFromAnonymous(it) == null }
+        otherSession.send(unknownIdentities)
+        progressTracker.currentStep = RECEIVING_PARTIES
+
+        val parties = otherSession.receive<List<Party>>().unwrap { it }
+        if (parties.isEmpty()) {
+            progressTracker.currentStep = NO_PARTIES_RECEIVED
+            return false
+        }
+        val mapConfidentialKeyToParty: Map<PublicKey, Party> = unknownIdentities.map { it.owningKey }.zip(parties).toMap()
```

Comment:
Am I right that you assume ordering on the list here? Please don't do that, if you want to have a mapping confidential identity -> well known, send such a mapping, don't zip two lists.

Reply:
That is correct. What's wrong with zip? I read somewhere on stackoverflow (classic) that is was fairly safe?